File: cpp_core/tests/test_market_data.cpp
#include "volarb/market_data.hpp"

#include <catch2/catch_test_macros.hpp>

#include <filesystem>
#include <string>

using volarb::ChainQuery;
using volarb::ContractQuote;
using volarb::CorruptDatasetError;
using volarb::format_canonical_date;
using volarb::format_canonical_timestamp;
using volarb::KnowledgeHorizon;
using volarb::LookaheadRequestedError;
using volarb::open_chain_dataset;
using volarb::parse_canonical_timestamp;

namespace {

const std::filesystem::path dataset_root = VOLARB_DATASET_ROOT;
const std::string correction_target = "SPX260918C04800000";
const std::string late_listing = "SPX260918C05520000";

std::vector<ContractQuote> chain(const std::string& horizon, const std::string& observation,
                                 bool include_adjusted = false) {
    const auto reader =
        open_chain_dataset(dataset_root, KnowledgeHorizon{parse_canonical_timestamp(horizon)});
    return reader.chain_as_of(ChainQuery{"SPX", parse_canonical_timestamp(observation), include_adjusted});
}

const ContractQuote* find(const std::vector<ContractQuote>& quotes, const std::string& symbol) {
    for (const ContractQuote& quote : quotes) {
        if (quote.contract_symbol == symbol) {
            return &quote;
        }
    }
    return nullptr;
}

}

TEST_CASE("canonical timestamps round trip", "[market_data]") {
    const std::string text = "2026-08-21T16:30:45.123456Z";
    REQUIRE(format_canonical_timestamp(parse_canonical_timestamp(text)) == text);
    REQUIRE(format_canonical_timestamp(0) == "1970-01-01T00:00:00.000000Z");
    REQUIRE(format_canonical_date(0) == "1970-01-01");
    REQUIRE(format_canonical_date(20686) == "2026-08-21");
}

TEST_CASE("a correction is invisible below its knowledge time", "[market_data]") {
    const auto before = chain("2026-08-21T17:00:00.000000Z", "2026-08-21T16:00:00.000000Z");
    const auto after = chain("2026-08-21T18:00:00.000000Z", "2026-08-21T16:00:00.000000Z");
    const ContractQuote* early = find(before, correction_target);
    const ContractQuote* late = find(after, correction_target);
    REQUIRE(early != nullptr);
    REQUIRE(late != nullptr);
    REQUIRE(early->bid_price != late->bid_price);
    REQUIRE(early->knowledge_time < late->knowledge_time);
}

TEST_CASE("a contract listed later is absent from earlier observations", "[market_data]") {
    REQUIRE(find(chain("2026-08-21T17:00:00.000000Z", "2026-08-21T16:00:00.000000Z"), late_listing) == nullptr);
    REQUIRE(find(chain("2026-08-21T17:00:00.000000Z", "2026-08-21T17:00:00.000000Z"), late_listing) != nullptr);
}

TEST_CASE("adjusted contracts are excluded unless requested", "[market_data]") {
    const auto standard_only = chain("2026-08-21T17:00:00.000000Z", "2026-08-21T17:00:00.000000Z");
    const auto with_adjusted = chain("2026-08-21T17:00:00.000000Z", "2026-08-21T17:00:00.000000Z", true);
    REQUIRE(with_adjusted.size() > standard_only.size());
    for (const ContractQuote& quote : standard_only) {
        REQUIRE(quote.is_standard_deliverable);
    }
}

TEST_CASE("raising the horizon never removes a contract", "[market_data]") {
    const auto narrow = chain("2026-08-21T17:00:00.000000Z", "2026-08-21T17:00:00.000000Z");
    const auto wide = chain("2026-08-22T12:00:00.000000Z", "2026-08-21T17:00:00.000000Z");
    for (const ContractQuote& quote : narrow) {
        REQUIRE(find(wide, quote.contract_symbol) != nullptr);
    }
}

TEST_CASE("lookahead is rejected rather than filtered", "[market_data]") {
    REQUIRE_THROWS_AS(chain("2026-08-21T16:00:00.000000Z", "2026-08-21T17:00:00.000000Z"),
                      LookaheadRequestedError);
}

TEST_CASE("an absent underlying is empty rather than an error", "[market_data]") {
    const auto reader = open_chain_dataset(
        dataset_root, KnowledgeHorizon{parse_canonical_timestamp("2026-08-21T17:00:00.000000Z")});
    const auto quotes =
        reader.chain_as_of(ChainQuery{"NVDA", parse_canonical_timestamp("2026-08-21T17:00:00.000000Z"), false});
    REQUIRE(quotes.empty());
}

TEST_CASE("a missing dataset is a typed error", "[market_data]") {
    REQUIRE_THROWS_AS(open_chain_dataset("/nonexistent/dataset", KnowledgeHorizon{0}), CorruptDatasetError);
}

TEST_CASE("results are sorted by expiry, strike, type, symbol", "[market_data]") {
    const auto quotes = chain("2026-08-21T17:00:00.000000Z", "2026-08-21T17:00:00.000000Z");
    REQUIRE(quotes.size() > 1);
    for (std::size_t index = 1; index < quotes.size(); ++index) {
        const ContractQuote& previous = quotes[index - 1];
        const ContractQuote& current = quotes[index];
        const bool ordered =
            std::tie(previous.expiry_date, previous.strike, previous.contract_symbol) <
            std::tie(current.expiry_date, current.strike, current.contract_symbol);
        REQUIRE(ordered);
    }
}

File: python_pure/tests/test_market_data.py
from __future__ import annotations

from datetime import UTC, datetime
from pathlib import Path

import pytest
from conftest import REPOSITORY_ROOT
from volarb_py.market_data import (
    ChainQuery,
    ContractQuote,
    CorruptDatasetError,
    KnowledgeHorizon,
    LookaheadRequestedError,
    chain_sort_key,
    format_canonical_date,
    format_canonical_timestamp,
    open_chain_dataset,
    parse_canonical_timestamp,
)

DATASET_ROOT = REPOSITORY_ROOT / "spec" / "fixtures" / "datasets" / "synthetic_chain"
CORRECTION_TARGET = "SPX260918C04800000"
SHA256_HEX_LENGTH = 64
LATE_LISTING = "SPX260918C05520000"


def moment(hour: int, minute: int = 0, day: int = 21) -> datetime:
    return datetime(2026, 8, day, hour, minute, tzinfo=UTC)


def chain(horizon: datetime, observation: datetime, include_adjusted: bool = False) -> list[ContractQuote]:
    reader = open_chain_dataset(DATASET_ROOT, KnowledgeHorizon(horizon))
    return reader.chain_as_of(ChainQuery("SPX", observation, include_adjusted))


def find(quotes: list[ContractQuote], symbol: str) -> ContractQuote | None:
    return next((quote for quote in quotes if quote.contract_symbol == symbol), None)


def test_canonical_timestamps_round_trip() -> None:
    text = "2026-08-21T16:30:45.123456Z"
    assert format_canonical_timestamp(parse_canonical_timestamp(text)) == text
    assert parse_canonical_timestamp("2026-08-21T16:30:45Z") == moment(16, 30).replace(second=45)
    assert format_canonical_date(datetime(2026, 8, 21).date()) == "2026-08-21"


def test_a_correction_is_invisible_below_its_knowledge_time() -> None:
    before = find(chain(moment(17), moment(16)), CORRECTION_TARGET)
    after = find(chain(moment(18), moment(16)), CORRECTION_TARGET)
    assert before is not None
    assert after is not None
    assert before.bid_price != after.bid_price
    assert before.knowledge_time < after.knowledge_time
    assert before.event_time == after.event_time


def test_a_later_listing_is_absent_from_earlier_observations() -> None:
    assert find(chain(moment(17), moment(16)), LATE_LISTING) is None
    assert find(chain(moment(17), moment(17)), LATE_LISTING) is not None


def test_adjusted_contracts_are_excluded_unless_requested() -> None:
    standard_only = chain(moment(17), moment(17))
    with_adjusted = chain(moment(17), moment(17), include_adjusted=True)
    assert len(with_adjusted) > len(standard_only)
    assert all(quote.is_standard_deliverable for quote in standard_only)
    assert any(not quote.is_standard_deliverable for quote in with_adjusted)


def test_raising_the_horizon_never_removes_a_contract() -> None:
    narrow = {quote.contract_symbol for quote in chain(moment(17), moment(17))}
    wide = {quote.contract_symbol for quote in chain(moment(12, day=22), moment(17))}
    assert narrow <= wide


def test_advancing_the_horizon_can_replace_a_row_with_a_newer_revision() -> None:
    same_day = find(chain(moment(17), moment(17)), CORRECTION_TARGET)
    next_morning = find(chain(moment(12, day=22), moment(17)), CORRECTION_TARGET)
    assert same_day is not None
    assert next_morning is not None
    assert next_morning.knowledge_time > same_day.knowledge_time
    assert next_morning.event_time == same_day.event_time


def test_no_returned_row_is_past_the_observation_time() -> None:
    for observation in (moment(15), moment(16), moment(17)):
        for quote in chain(moment(12, day=22), observation):
            assert quote.event_time <= observation


def test_no_returned_row_is_past_the_knowledge_horizon() -> None:
    for horizon in (moment(15), moment(17), moment(18), moment(12, day=22)):
        for quote in chain(horizon, moment(15)):
            assert quote.knowledge_time <= horizon


def test_lookahead_is_rejected_rather_than_filtered() -> None:
    with pytest.raises(LookaheadRequestedError, match="past knowledge horizon"):
        chain(moment(16), moment(17))


def test_an_absent_underlying_is_empty_rather_than_an_error() -> None:
    reader = open_chain_dataset(DATASET_ROOT, KnowledgeHorizon(moment(17)))
    assert reader.chain_as_of(ChainQuery("NVDA", moment(17))) == []


def test_a_missing_dataset_is_a_typed_error(tmp_path: Path) -> None:
    with pytest.raises(CorruptDatasetError, match=r"no manifest\.json"):
        open_chain_dataset(tmp_path / "absent", KnowledgeHorizon(moment(17)))


def test_results_are_totally_ordered() -> None:
    quotes = chain(moment(17), moment(17))
    keys = [chain_sort_key(quote) for quote in quotes]
    assert keys == sorted(keys)
    assert len(set(keys)) == len(keys)


def test_the_reader_exposes_the_dataset_digest_for_pinning() -> None:
    reader = open_chain_dataset(DATASET_ROOT, KnowledgeHorizon(moment(17)))
    assert len(reader.dataset_digest) == SHA256_HEX_LENGTH
    assert reader.knowledge_horizon == moment(17)


def test_the_reader_offers_no_way_to_bypass_the_horizon() -> None:
    reader = open_chain_dataset(DATASET_ROOT, KnowledgeHorizon(moment(17)))
    public_operations = {name for name in dir(reader) if not name.startswith("_")}
    assert public_operations == {"chain_as_of", "dataset_digest", "knowledge_horizon"}
